framework/Optimizers/stepManipulators/ConjugateGradient.py

```diff
@@ -0,0 +1,436 @@
+import abc
+import numpy as np
+from scipy.optimize import minpack2
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from utils import utils, InputData, InputTypes, mathUtils, randomUtils
+from .StepManipulator import StepManipulator
+from . import NoConstraintResolutionFound, NoMoreStepsNeeded
+#Internal Modules End--------------------------------------------------------------------------------
+
+class ConjugateGradient(StepManipulator):
+  """
+    Changes step size depending on history of gradients
+  """
+  requiredInformation = ['gradientHist', 'prevStepSize']
+  optionalInformation = ['recommend']
+
+  ##########################
+  # Initialization Methods #
+  ##########################
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for class cls.
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for specifying input of cls.
+    """
+    specs = super(ConjugateGradient, cls).getInputSpecification()
+    return specs
+
+  @classmethod
+  def getSolutionExportVariableNames(cls):
+    """
+      Compiles a list of acceptable SolutionExport variable options.
+      @ In, None
+      @ Out, vars, list(str), list of acceptable variable names
+    """
+    ok = super(ConjugateGradient, cls).getSolutionExportVariableNames()
+    ok['CG_task'] = 'for ConjugateGradient, current task of line search. FD suggests continuing the search, and CONV indicates the line search converged and will pivot.'
+    return ok
+
+  def __init__(self):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    StepManipulator.__init__(self)
+    ## Instance Variable Initialization
+    # public
+    self.needsAccessToAcceptance = True # if True, then this stepManip may need to modify opt point acceptance criteria
+    # _protected
+    self._persistence = None     # consecutive line search converges until acceptance
+    # __private
+    # additional methods
+
+  def handleInput(self, specs):
+    """
+      Read input specs
+      @ In, specs, InputData.ParameterInput, parameter specs interpreted
+      @ Out, None
+    """
+    #specs = specs
+    growth = specs.findFirst('growthFactor')
+    if growth is not None:
+      self._growth = growth.value
+    shrink = specs.findFirst('shrinkFactor')
+    if shrink is not None:
+      self._shrink = shrink.value
+
+  def initialize(self, optVars, persistence=None, **kwargs):
+    """
+      initializes this object
+      @ In, optVars, list(str), list of optimization variables (e.g. input space)
+      @ In, persistence, integer, successive converges required to consider total convergence
+      @ In, kwargs, dict, additional unused arguments
```

Comment:
unused? why do we have them?

Reply:
because some classes have other keywords that they do use, that ConjugateGradient does not, or vice versa. Without this, you need a custom calling format for each subclass.